File: src/main/java/org/example/Service/TarefaService.java
package org.example.Service;

import org.example.Repository.TarefaRepository;
import org.example.Repository.UsuarioRepository;
import org.example.model.Tarefa;
import org.example.model.Usuario;
import org.springframework.stereotype.Service;

import java.util.ArrayList;
import java.util.List;

@Service
public class TarefaService {
    private final TarefaRepository tarefaRepository;
    private final UsuarioRepository usuarioRepository;
    public TarefaService(TarefaRepository tarefaRepository, UsuarioRepository usuarioRepository) {
        this.tarefaRepository = tarefaRepository;
        this.usuarioRepository = usuarioRepository;
    }

    public Tarefa salvar(Tarefa tarefa) {

        if (tarefa.getAdministrador() != null && tarefa.getAdministrador().getId() != null) {
            Usuario admin = usuarioRepository.findById(tarefa.getAdministrador().getId())
                    .orElseThrow(() -> new RuntimeException("Administrador não encontrado"));
            tarefa.setAdministrador(admin);
        }

        if (tarefa.getMembros() == null) {
            tarefa.setMembros(new ArrayList<>());
        } else {
            List<Usuario> membrosCarregados = tarefa.getMembros().stream()
                    .map(membro -> usuarioRepository.findById(membro.getId())
                            .orElseThrow(() -> new RuntimeException("Membro ID " + membro.getId() + " não encontrado")))
                    .toList();
            tarefa.setMembros(new ArrayList<>(membrosCarregados));
        }

        if (tarefa.getAdministrador() != null && !tarefa.getMembros().contains(tarefa.getAdministrador())) {
            tarefa.getMembros().add(tarefa.getAdministrador());
        }

        return tarefaRepository.save(tarefa);
    }

    public Tarefa findById(Long id){
        return tarefaRepository.findById(id).orElse(null);
    }

    public Tarefa atualizar(Tarefa tarefa) {
        return tarefaRepository.save(tarefa);
    }

    public List<Tarefa> listarPorUsuario(Long usuarioId) {
        return tarefaRepository.findByUsuarioEnvolvido(usuarioId);
    }
}

File: gerenciador_tarefas/lib/services/tarefa_service.dart
import '../models/tarefa.dart';
import 'api_service.dart';

class TarefaService {
  static Future<Tarefa> buscarPorId(int id) async {
    final data = await ApiService.get('/Tarefas/$id');
    return Tarefa.fromJson(data);
  }

  static Future<List<Tarefa>> listarPorUsuario(int usuarioId) async {
    final data = await ApiService.get('/Tarefas/usuario/$usuarioId');
    return (data as List).map((e) => Tarefa.fromJson(e)).toList();
  }

  static Future<Tarefa> salvar(Tarefa tarefa) async {
    final data = await ApiService.post('/Tarefas', tarefa.toJson());
    return Tarefa.fromJson(data);
  }

  static Future<Tarefa> atualizar(Tarefa tarefa) async {
    final data = await ApiService.patch('/Tarefas', tarefa.toJson());
    return Tarefa.fromJson(data);
  }
}